File: Feel@Home/Feel@Home/src/main/java/club/frickel/feelathome/OnlyOneDelayEffectConfigFragment.java
/*
 * Copyright (C) 2012 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package club.frickel.feelathome;

import android.app.Fragment;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;


public class OnlyOneDelayEffectConfigFragment extends Fragment implements TextWatcher {

    private String delay = "50";

    private EditText delayTextEdit;

    private String deviceID;
    Effect effect;
    Map<String, Object> effectConfig;

    public OnlyOneDelayEffectConfigFragment(){}

    private void sendState() {
        if (effectConfig != null){
            effect.setConfig(effectConfig);
        }
        new SendStateHandler(effect, deviceID, getActivity()).execute();
    }

    private void updateOneDelayEffectView() {

        effectConfig = new HashMap<>();
        if (effect.getConfig() != null) {
            effectConfig = effect.getConfig();
            String delayString = effectConfig.get("Delay").toString();
            Pattern pattern = Pattern.compile("(-)?(\\d+)");
            Matcher matcher = pattern.matcher(delayString);
            if (matcher.find()) {
                delay =  matcher.group(0);
            }
        }
        delayTextEdit.setText(delay);

    }


    @Override
    public void onCreate (Bundle savedInstanceState){
        super.onCreate(savedInstanceState);
        Bundle arguments = getArguments();
        if (arguments != null){
            effect = (Effect)arguments.getSerializable(Constants.EFFECT_STRING);
            deviceID = arguments.getString(Constants.DEVICE_ID);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        super.onCreateView(inflater,container,savedInstanceState);
        if(savedInstanceState != null){
            this.deviceID = savedInstanceState.getString(Constants.DEVICE_ID);
            this.effect = (Effect)savedInstanceState.getSerializable(Constants.EFFECT_STRING);
        }

        final View view = inflater.inflate(R.layout.only_one_delay_config_layout, container, false);
        delayTextEdit = (EditText) view.findViewById(R.id.effect_config_delay_editText);

        updateOneDelayEffectView();

        return view;
    }

    @Override
    public void onStart() {
        super.onStart();
        delayTextEdit.addTextChangedListener(this);
    }

    @Override
    public void onStop() {
        super.onStop();
        delayTextEdit.removeTextChangedListener(this);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        outState.putString(Constants.DEVICE_ID, deviceID);
        outState.putSerializable(Constants.EFFECT_STRING,effect);
        super.onSaveInstanceState(outState);
    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        //To change body of implemented methods use File | Settings | File Templates.
    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
        delay = s.toString();
        effectConfig.put("Delay", delay + "ms");
        sendState();
    }

    @Override
    public void afterTextChanged(Editable s) {

    }
}

File: Feel@Home/Feel@Home/src/main/java/club/frickel/feelathome/OnlyOneDelayEffectConfigFragment.kt
package club.frickel.feelathome

import android.app.Fragment
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import java.util.*
import java.util.regex.Pattern


class OnlyOneDelayEffectConfigFragment : Fragment(), TextWatcher {

    private var delay = "50"

    private lateinit var delayTextEdit: EditText

    private var deviceID: String? = null

    internal var effect: Effect = Effect("", "", emptyMap())
    internal var effectConfig: MutableMap<String, Any> = emptyMap<String, Any>().toMutableMap()

    private fun sendState() {
        if (effectConfig != null) {
            effect.config = effectConfig
        }
        if (effect != null) SendStateHandler(effect, deviceID, activity).execute()
    }

    private fun updateOneDelayEffectView() {

        effectConfig = HashMap<String, Any>()
        if (effect.config != null) {
            effectConfig = effect.config.toMutableMap()
            val delayString = effectConfig!!["Delay"].toString()
            val pattern = Pattern.compile("(-)?(\\d+)")
            val matcher = pattern.matcher(delayString)
            if (matcher.find()) {
                delay = matcher.group(0)
            }
        }
        delayTextEdit.setText(delay)

    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val arguments = arguments
        if (arguments != null) {
            effect = arguments.getSerializable(Constants.EFFECT_STRING) as Effect
            deviceID = arguments.getString(Constants.DEVICE_ID)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        if (savedInstanceState != null) {
            this.deviceID = savedInstanceState.getString(Constants.DEVICE_ID)
            this.effect = savedInstanceState.getSerializable(Constants.EFFECT_STRING) as Effect
        }

        val view = inflater.inflate(R.layout.only_one_delay_config_layout, container, false)
        delayTextEdit = view.findViewById(R.id.effect_config_delay_editText) as EditText

        updateOneDelayEffectView()

        return view
    }

    override fun onStart() {
        super.onStart()
        delayTextEdit.addTextChangedListener(this)
    }

    override fun onStop() {
        super.onStop()
        delayTextEdit.removeTextChangedListener(this)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putString(Constants.DEVICE_ID, deviceID)
        outState.putSerializable(Constants.EFFECT_STRING, effect)
        super.onSaveInstanceState(outState)
    }

    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
        //To change body of implemented methods use File | Settings | File Templates.
    }

    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
        delay = s.toString()
        effectConfig.put("Delay", delay + "ms")
        sendState()
    }

    override fun afterTextChanged(s: Editable) {

    }
}